substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1980,6 +1983,115 @@ pub mod pallet {
+		///
+		/// The `maybe_*` input parameters will overwrite the corresponding data and metadata of the
+		/// ledger associated with the stash. If the input parameters are not set, the ledger will
+		/// be reset with the current values.
+		///
+		/// Upon successful execution, this extrinsic will *recreate* the ledger based on its past
+		/// state and/or the `maybe_controller` and `maybe_total` input parameters.
+		#[pallet::call_index(29)]
+		#[pallet::weight(T::WeightInfo::restore_ledger())]
+		pub fn restore_ledger(
+			origin: OriginFor<T>,
+			stash: T::AccountId,
+			maybe_controller: Option<T::AccountId>,
+			maybe_total: Option<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			// closure to call before return from extrinsic to ensure the restore was successful.
+			let integrity_checks = || {
+				let lock = T::Currency::balance_locked(crate::STAKING_ID, &stash);
+
+				if let Some(ledger) = Bonded::<T>::get(&stash).and_then(Ledger::<T>::get) {
+					ensure!(lock == ledger.total, Error::<T>::BadState);
+					ensure!(Payee::<T>::get(&stash).is_some(), Error::<T>::BadState);
+					ensure!(ledger.stash == stash, Error::<T>::BadState);
+				} else {
+					ensure!(lock == Zero::zero(), Error::<T>::BadState);
+				}
+
+				Ok::<_, Error<T>>(())
+			};
+
+			let current_lock = T::Currency::balance_locked(crate::STAKING_ID, &stash);
+
+			let (new_controller, new_total) = match Self::inspect_bond_state(&stash) {
+				Ok(LedgerIntegrityState::Corrupted) => {
+					let new_controller = maybe_controller.unwrap_or(stash.clone());
+
+					let new_total = if let Some(new_total) = maybe_total {
+						// enforce lock == ledger.amount.
+						T::Currency::set_lock(
+							crate::STAKING_ID,
+							&stash,
+							new_total,
+							WithdrawReasons::all(),
+						);
+						new_total
+					} else {
+						current_lock
+					};
+
+					Ok((new_controller, new_total))
+				},
+				Ok(LedgerIntegrityState::CorruptedKilled) => {
+					if current_lock == Zero::zero() {
+						// this case needs to restore both lock and ledger, so the new total needs
+						// to be given by the called since there's no way to restore the total
+						// on-chain.
+						ensure!(maybe_total.is_some(), Error::<T>::CannotResetLedger);
+						Ok((
+							stash.clone(),
+							maybe_total.expect("total exists as per the check above; qed."),
+						))
+					} else {
+						Ok((stash.clone(), current_lock))
+					}
+				},
+				Ok(LedgerIntegrityState::LockCorrupted) => {
+					// ledger is not corrupted but its locks are out of sync. In this case, we need
+					// to enforce a new ledger.total and staking lock for this stash.
+					let new_total = maybe_total.ok_or(Error::<T>::CannotResetLedger)?;
+					T::Currency::set_lock(
+						crate::STAKING_ID,
+						&stash,
+						new_total,
+						WithdrawReasons::all(),
+					);
+
+					Ok((stash.clone(), new_total))
+				},
+				Err(Error::<T>::BadState) => {
```

Comment:
If `stash` is a random account in staking, we return `BadState` and come to this arm right, and this arm provides free execution, which means I can endlessly spam the chain?

Reply:
oh well, `AdminOrigin`, nvmd. But good for you to be aware of this detail.